crossbar/worker/proxy.py

```diff
@@ -0,0 +1,537 @@
+        message.Welcome,
+        message.Authenticate,
+    )
+
+    def set_backend(self, other_session):
+        self._backend = other_session
+
+    def onMessage(self, msg):
+        if isinstance(msg, self.do_not_forward):
+            return super(ProxyServerSession, self).onMessage(msg)
+        # TODO: should always destroy 'the other side' if one side
+        # (including the backend) goes away (e.g. message IDs would
+        # get out of sync if we just auto re-connected to the backend)
+        if not self._backend or not self._backend._transport:
+            raise RuntimeError("No backend, or backend transport gone")
+        self._backend._transport.send(msg)
+
+
+class ProxyRealm(object):
+    """
+    Faking out some things so we can provide a valid _realm object for
+    ProxyRouter
+    """
+
+    def __init__(self, realm_id, session):
+        self._realm_id = realm_id
+        self.session = session  # the authenticator (among others?) will call on this session
+
+
+class ProxyRouter(object):
+    """
+    Provide enough Router functionality so that crossbar's base
+    RouterSession class can call various methods on us..
+    """
+    log = txaio.make_logger()
+
+    def __init__(self, realm, worker, transport_config):
+        self._realm = ProxyRealm(realm, worker)
+        self._worker = worker
+        self._front_to_back = dict()  # session_id -> ProxySession backend
+        self._backend_transport_config = transport_config
+
+    def has_role(self, role):
+        """
+        we just forward any role at all onto the 'real' backend
+
+        XXX TODO: is this true? or should we config which roles are
+        okay? or is that up to 'the real backend' itself?
+        """
+        return True
+
+    def _make_backend_connection(self, realm, authid, authrole):
+        backend_connected_d = Deferred()
+
+        self.log.info(
+            "backend connection to: {config}",
+            config=self._backend_transport_config,
+        )
+        ep = create_connecting_endpoint_from_config(
+            self._backend_transport_config['endpoint'],
+            self._worker._cbdir,
+            self._worker._reactor,
+            self.log,
+        )
+
+        def create_session():
+            self.log.info("ProxyRouter connected, creating session", )
+            # backend_cfg = list(self._worker._backend_configs.values())[0]
+            s = Session(types.ComponentConfig(realm=realm))
+            s.add_authenticator(AuthProxy(
+                authrealm=realm,
+                authid=authid,
+                authrole=authrole,
+            ))
+
+            def _join(*args):
+                self.log.info("ProxyRouter joined to backend session", )
+                backend_connected_d.callback(s)
+
+            s.on('join', _join)
+            return s
+
+        # XXX FIXME TODO look at transport-type etc
+        factory = WampWebSocketClientFactory(create_session, url="ws://localhost:9000/ws")
```

Comment:
Yeah, this means we are just hard-coding "websocket" as the only proxy-able connection; need to handle others via config.